internal/directives/argocd_update_directive.go

```diff
@@ -0,0 +1,1104 @@
+) error {
+	// Initiate a "hard" refresh.
+	if app.ObjectMeta.Annotations == nil {
+		app.ObjectMeta.Annotations = make(map[string]string, 1)
+	}
+	app.ObjectMeta.Annotations[argocd.AnnotationKeyRefresh] = string(argocd.RefreshTypeHard)
+
+	// Update the desired source(s) in the Argo CD Application.
+	app.Spec.Source = desiredSource.DeepCopy()
+	app.Spec.Sources = desiredSources.DeepCopy()
+
+	// Initiate a new operation.
+	app.Operation = &argocd.Operation{
+		InitiatedBy: argocd.OperationInitiator{
+			Username:  applicationOperationInitiator,
+			Automated: true,
+		},
+		Info: []*argocd.Info{
+			{
+				Name:  "Reason",
+				Value: "Promotion triggered a sync of this Application resource.",
+			},
+			{
+				Name:  freightCollectionInfoKey,
+				Value: stepCtx.Freight.ID,
+			},
+		},
+		Sync: &argocd.SyncOperation{
+			Revisions: []string{},
+		},
+	}
+	if app.Spec.SyncPolicy != nil {
+		if app.Spec.SyncPolicy.Retry != nil {
+			app.Operation.Retry = *app.Spec.SyncPolicy.Retry
+		}
+		if app.Spec.SyncPolicy.SyncOptions != nil {
+			app.Operation.Sync.SyncOptions = app.Spec.SyncPolicy.SyncOptions
+		}
+	}
+	if app.Spec.Source != nil {
+		app.Operation.Sync.Revisions = []string{app.Spec.Source.TargetRevision}
+	}
+	for _, source := range app.Spec.Sources {
+		app.Operation.Sync.Revisions = append(app.Operation.Sync.Revisions, source.TargetRevision)
+	}
+
+	// Patch the Argo CD Application.
+	if err := a.argoCDAppPatchFn(ctx, stepCtx, app, func(src, dst unstructured.Unstructured) error {
+		// If the resource has been modified since we fetched it, an update
+		// can result in unexpected merge results. Detect this, and return an
+		// error if it occurs.
+		if src.GetGeneration() != dst.GetGeneration() {
+			return fmt.Errorf("unable to update sources to desired revisions: resource has been modified")
+		}
+
+		dst.SetAnnotations(src.GetAnnotations())
+		dst.Object["spec"] = recursiveMerge(src.Object["spec"], dst.Object["spec"])
+		dst.Object["operation"] = src.Object["operation"]
+		return nil
+	}); err != nil {
+		return fmt.Errorf("error patching Argo CD Application %q: %w", app.Name, err)
+	}
+	logging.LoggerFromContext(ctx).Debug("patched Argo CD Application", "app", app.Name)
+
+	// NB: This attempts to mimic the behavior of the Argo CD API server,
+	// which logs an event when a sync is initiated. However, we do not
+	// have access to the same enriched event data the Argo CD API server
+	// has, so we are limited to logging an event with the best
+	// information we have at hand.
+	// xref: https://github.com/argoproj/argo-cd/blob/44894e9e438bca5adccf58d2f904adc63365805c/server/application/application.go#L1887-L1895
+	// nolint:lll
+	//
+	// TODO(hidde): It is not clear what we should do if we have a list of
+	// sources.
+	message := "initiated sync"
+	if app.Spec.Source != nil {
+		message += " to " + app.Spec.Source.TargetRevision
+	}
+	a.logAppEventFn(
+		ctx,
+		stepCtx,
+		app,
+		"kargo-controller",
```

Comment:
Appears like this could make use of `applicationOperationInitiator`?

Reply:
Fixes in 18dabd66be1c30110ced578cbf5ea2a54e55d1b1